### PR TITLE
[Build] Support Debian snapshot mirror to improve build stability

### PR DESCRIPTION
**Pull request by xumia**
#### Why I did it
[Build] Support Debian snapshot mirror to improve build stability

It is to enhance the reproducible build, supports the Debian snapshot mirror. It guarantees all the docker images using the same Debian mirror snapshot and fixes the temporary build failure which is caused by remote Debain mirror indexes changed during the build. It is also to fix the version conflict issue caused by no fixed versions of some of the Debian packages.

#### How I did it
Add a new feature to support the Debian snapshot mirror.

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106
- [ ] 202111
- [ ] 202205
- [ ] 202211

### CODE REVIEW

File: src/sonic-build-hooks/scripts/buildinfo_base.sh [hunk 1]
```diff
@@ -111,14 +111,37 @@ set_reproducible_mirrors()
 {
     # Remove the charater # in front of the line if matched
     local expression="s/^#\(.*$REPR_MIRROR_URL_PATTERN\)/\1/"
+    local expression2="/^#*deb.*$REPR_MIRROR_URL_PATTERN/! s/^#*deb/#&/"
```
Comment (qiluo-msft) [inline src/sonic-build-hooks/scripts/buildinfo_base.sh:114]: >expression2 [](http://example.com/codeflow?start=10&length=11)

Need code comment for each regex.

File: src/sonic-build-hooks/scripts/buildinfo_base.sh [hunk 2]
```diff
@@ -111,14 +111,37 @@ set_reproducible_mirrors()
...
+    local expression3="\$a#SET_REPR_MIRRORS"
     if [ "$1" = "-d" ]; then
         # Add the charater # in front of the line if match
         expression="s/^deb.*$REPR_MIRROR_URL_PATTERN/#\0/"
+        expression2="/^#*deb.*$REPR_MIRROR_URL_PATTERN/! s/^#(#*deb)/\1/"
```
Comment (qiluo-msft) [inline src/sonic-build-hooks/scripts/buildinfo_base.sh:119]: >expression2 [](http://example.com/codeflow?start=8&length=11)

The same.

File: src/sonic-build-hooks/scripts/collect_version_files [hunk 3]
```diff
@@ -25,6 +25,13 @@ dpkg-query -W -f '${Package}==${Version}\n' | grep -Ev "${SKIP_VERSION_PACKAGE}"
 ## Add the the packages purged
 [ -f $POST_VERSION_PATH/purge-versions-deb ] && cat $POST_VERSION_PATH/purge-versions-deb >> "${TARGET_PATH}/versions-deb-${DIST}-${ARCH}"
 
+## Add mirror versions
+while read -r line; do
+  mirror=$(echo "$line" | sed "s/.*\///" | sed "s/_InRelease.*//")
```
Comment (qiluo-msft) [inline src/sonic-build-hooks/scripts/collect_version_files:30]: >   [](http://example.com/codeflow?start=0&length=2)

inconsistent indentation.

File: scripts/generate_buildinfo_config.sh [hunk 4]
```diff
@@ -11,3 +11,4 @@ echo "export SONIC_VERSION_CONTROL_COMPONENTS=$SONIC_VERSION_CONTROL_COMPONENTS"
 echo "export SONIC_VERSION_CACHE=${SONIC_VERSION_CACHE}" >> $BUILDINFO_CONFIG
 echo "export SONIC_VERSION_CACHE_SOURCE=${SONIC_VERSION_CACHE_SOURCE}" >> $BUILDINFO_CONFIG
 echo "export DISTRO=${DISTRO}" >> $BUILDINFO_CONFIG
+echo "export MIRROR_SNAPSHOT=$MIRROR_SNAPSHOT"  >> $BUILDINFO_CONFIG
```
Comment (qiluo-msft) [inline scripts/generate_buildinfo_config.sh:14]: >  [](http://example.com/codeflow?start=47&length=1)

remove one extra blank.

File: src/sonic-build-hooks/scripts/buildinfo_base.sh (hunk 2 above)
Comment (xumia) [reply src/sonic-build-hooks/scripts/buildinfo_base.sh:122]: Added the comment, thanks.

File: src/sonic-build-hooks/scripts/buildinfo_base.sh (hunk 1 above)
Comment (xumia) [reply src/sonic-build-hooks/scripts/buildinfo_base.sh:116]: Added the comment, thanks.

File: src/sonic-build-hooks/scripts/collect_version_files (hunk 3 above)
Comment (xumia) [reply src/sonic-build-hooks/scripts/collect_version_files]: Fixed it, thanks.

File: scripts/generate_buildinfo_config.sh (hunk 4 above)
Comment (xumia) [reply scripts/generate_buildinfo_config.sh]: Fixed it, thanks.

**Review by qiluo-msft — APPROVED** [review]

**Review by liushilongbuaa — APPROVED** [review]

**Comment by xumia** [comment]
> @xumia
> 
> You don't `export MIRROR_SNAPSHOT` in Makefile.work, so you don't pass this flag to build_mirror_config.sh when generate sources.list for sonic-slaves. Is it a correct behaviour?

@k-v1, I think the variable should be exported. Looks like it will not work if passing the variable in the make variables.

**Comment by xumia** [comment]
> http://debian-archive.trafficmanager.net/debian

Good catch, we should use the snapshot one, when the flag set.

**Pull request merged**